APID: Doc about migrating from MS1 to MS2 (#97)
mulesoft/docs-design-center: modules/ROOT/pages/design-migrating-ms.adoc

> = Migrate to the Version 2 of the Mocking Service

Reviewer: Should we strike the "the" here to be consistent with the subtitle below?

> How to Migrate an API-Specification Project to Version 2 of the Mocking Service

Author: Thanks.

The mocking service lets you test endpoints that are defined in an API specification. The advantage to using the mocking service is that you do not need to implement the API before testing. To test an endpoint, the mocking service simulates calls to its methods.

For example, when you are designing an API specification in API Designer, you can simulate a call to an endpoint’s GET method to see whether the response body for the 200 status code is as you want it. As a further example, if you are evaluating whether you want to implement an API specification that is in Exchange, you can simulate a call to an endpoint’s GET method to find out what the response body looks like.

[IMPORTANT]
====

> After January 31, 2020, version 1 of the mocking service will no longer send responses to requests. If you are using a third-party tool to send requests, the mocking service will return the HTTP status code 302 and provide a link to this document.

Author: Good question about the date. It's in the FAQ doc, but that is internal.
@harshakirank , what do you think?

As for the 302, it's in the internal FAQ here: https://docs.google.com/document/d/1haH2AxX15SNx8ZpjvFKZqGWobpEcfqQKsgeq4YJ0JdE/edit#heading=h.3c4p43jwicko
Perhaps it's a typo in the FAQ? @harshakirank

> After January 31, 2020, version 1 of the mocking service will no longer send responses to requests. If you are using a third-party tool to send requests, the mocking service will return the HTTP status code 302 and provide a link to this document.

Reviewer 2: Yes, the date is fine, and my bad, the response code has to be 301.

> After January 31, 2020, version 1 of the mocking service will no longer send responses to requests. If you are using a third-party tool to send requests, the mocking service will return the HTTP status code 302 and provide a link to this document.

Reviewer 2: Please update the status code to 301

> After January 31, 2020, version 1 of the mocking service will no longer send responses to requests. If you are using a third-party tool to send requests, the mocking service will return the HTTP status code 302 and provide a link to this document.

Reviewer 2: Lets the service deprecation date Nov 31 and explain that critical users can still access the Mocking service 1.x via passing this header x- (let's add the exact header info from Martin), and on Jan 31, 2020, the Mocking Service will be completely decommissioned.

Reviewer 3: The header is not yet defined, so to avoid confusion to the user at this point I would say "via passing a special header". Then, in Nov we can update the doc again with the specific instructions and the header already defined

Author: @mnbattaglia and @harshakirank , the problem is that users will need the header before November 31, so that there is no break in their continuous use of third-party testing tools. So, I'll need to include a date on which I will update this doc with the header.

Reviewer 3: Fair enough. I'll have that by today EOD

====

== How to Migrate an API-Specification Project to Version 2 of the Mocking Service

> If an API-specification project is older than January 10, 2019 and the mocking service was switched on before that date, follow these steps to migrate to version 2 of the mocking service:

Reviewer: I think this sentence is a little too verbose, and the "…switched on before" portion was slightly confusing to me. Maybe something like:

> If the mocking service was switched on for a project before January 10, 2019, follow these steps to migrate to version 2:

Reviewer 3: I agree with Tony. Also, it doesn matter if the project is older than January 10, 2019. That's redundant

Author: Good suggestion. Thanks.

> If an API-specification project is older than January 10, 2019 and the mocking service has been switched on since that date, follow these steps to migrate to version 2 of the mocking service:

Author: Right, I mangled that. Fixed.

> . In the notification that appears in the top-right corner after you open the project, click *Upgrade*. The text of the notification reads “A new mocking service is available. Upgrading takes just seconds.”

Reviewer: I think it's not necessary to quote the text of the notification. If the docs want to convey the same message, it should do so naturally.

Author: Yeah, I was on the fence with that, thinking I needed to add some way of identifying the notification, so that users could feel certain that they were clicking the right one. I know that there is only one, but users won't necessarily know that there is only one notification in APID. They might assume that there could be others for other purposes.
Removed.

+
*Result:* The mocking service is switched off and then on. The format of the base URI that the mocking service generates now differs from the format of the previous base URI.
+
Version 1 of the mocking service used base URIs that began with `https://mocksvc.mulesoft.com`.
+
Version 2 uses base URIs that begin with `https://anypoint.mulesoft.com/mocking/api/v1/links/`.

> . If any third-party tools are testing endpoints in the API specification for that project, replace the previous base URI with the new base URI in the links that those tools are using.

Reviewer 4: The link id will also change after switching off/on mocking service, so the user not only has to change the base URI but need to change the link id as well

Author: @dfmacias So, in version 1, the link followed directly after https://mocksvc.mulesoft.com/?

Reviewer 4: In version 1, a mock URL looks like this:
`https://mocksvc.mulesoft.com/mocks/473749b1-8b67-44bd-a572-3b362b34fb89`

after migrating this mock from APID, the URL will look like this:
`https://anypoint.mulesoft.com/mocking/api/v1/links/512f2bef-3efd-4678-bf43-9f96abab65b2/`

so, the user should replace the first URL with the second one, as the mock id changes.

What do you mean when saying `base URI`? do you mean the whole URL as shown in the examples above? If that the case, it is correct

Author: @dfmacias , by "base URI" I mean the URI that the mocking service appends to when generating the URIs for methods.
For example, this is the base URI for a spec:
ttps://anypoint.mulesoft.com/mocking/api/v1/links/7d19e624-a0d6-4d77-9fb8-8c78a8aeb575/api
For this particular GET method on the /customers endpoint, this is the URI:
https://anypoint.mulesoft.com/mocking/api/v1/links/7d19e624-a0d6-4d77-9fb8-8c78a8aeb575/api/customers

Customers really need to replace only the base URI, not the full URI.

Reviewer 4: I totally agree. I just made the comment because it was not clear to me what you meant with `base URI`

it is clearer now with the example. Just one last comment, the link id will change, so if you had this link

`https://mocksvc.mulesoft.com/mocks/7d19e624-a0d6-4d77-9fb8-8c78a8aeb575/api/customers`

the id `7d19e624-a0d6-4d77-9fb8-8c78a8aeb575` will be changed with a new one, for instance

`https://anypoint.mulesoft.com/mocking/api/v1/links/512f2bef-3efd-4678-bf43-9f96abab65b2/api/customers`

Author: Got it. I've changed the ID from MS2 in the example that I added to the doc.

== New Features in Version 2 That You Can Use in API Designer

Version 2 of the mocking service, which was released on January 10, 2019, introduced the following features that you can use in API Designer. 

=== Support for OAS 2.0

You can now test endpoints that are in API specifications that are written in OAS 2.0.

> === Behavioral headers

Reviewer: Should this header (and the one below) be title case?

Author: I used to use sentence case everywhere, but Kim has me using title case now. So, yeah.

> Version 2 of the mocking service, which was released on January 10, 2019, introduced the following list of features that you can use in API Designer. For a full list of features, see the https://anypoint.mulesoft.com/exchange/org.mule.examples/mocking-service-2-api/version/v1/pages/Getting%20Started/[Getting Started page for the Mocking Service 2 API]. (The link requires that you authenticate to Anypoint Platform.)

Reviewer 4: This link points to an Exchange Portal that explains how to use the Mocking Service 2 API. Not sure if that page will help or not the users understand the benefit of migrating to version 2 of mocking service

Author: Removed.

Add headers to your simulated calls to methods, changing the behavior of the mocking service in response to those calls The following new headers are available. See xref::apid-behavioral-headers.adoc[Add Behavioral Headers to Simulated API Calls] for full details.

* MS2-Delay: Indicates how many milliseconds the request should take to answer.
* MS2-Error-Level: Indicates which range of error status codes the mocking service is restricted to use.
* MS2-Error-Rate: Indicates the probability of getting an error status code as a response.
* MS2-Randomize: When this header is set to true, the mocking service auto-generates a payload response if no example is defined for the response.
* MS2-Status-Code: Specifies the status code or codes that the mocking service can choose from for the return message.
* MS2-Strict-Model-Validations: When you set this header to false, you can use the mocking service even when the API specification contains errors

=== Performance improvements

> The mocking service uses fewer resources, but responds to requests in less time on average.

Reviewer: I'm a little confused by the usage of "but" here. Is using fewer resources a negative outcome of migrating?

Author: Nice catch.

> === Authenticated Access to API Specifications That are Shared on Exchange

Reviewer: Should "are" be capitalized as well?

Author: I always seem to neglect capitalizing the conjugated forms of "to be".

Only Anypoint Platform users can simulate calls to your API specification after it is published to a private instance of Exchange, whether they are simulating calls in Exchange or using third-party tools. The mocking service now uses a base URI in Exchange that differs from the one that you can test with in API Designer.

> If a person is in Exchange and trying to simulate a call to a method, that person is prompted in API Designer to supply OAuth 2.0 credentials before sending a request.

Reviewer 3: I think the API Designer prompt part is wrong. If a person is in Exchange and use the console to simulate the call, it will just work (as Exchange will send the Authorization header under the hood). The thing is with a third party tool as it is explained in the following paragraph

Author: Yes, I mangled that, too. Thanks.

> If a person is using a third-party tool to simulate a call, that person first needs to send a `POST` request that sends the username and password for that person’s Anypoint Platform account.

Reviewer 3: I would say something like:
If a person is using a third-party tool to simulate a call to an Exchange asset, an authorization header needs to be sent. That person first needs to send a `POST` request that sends the username and password for that person’s Anypoint Platform account.

= Migrate to Version 2 of the Mocking Service

> On November 31, 2019, version 1 of the mocking service will be deprecated. Starting on that date, third-party tools will need to include this header in requests that they send to the mocking service to test endpoints in API Designer:

Reviewer 2: The November 31 is not a real date we need to update it to November 29(Friday).

> On November 31, 2019, version 1 of the mocking service will be deprecated. Starting on that date, third-party tools will need to include this header in requests that they send to the mocking service to test endpoints in API Designer:

Reviewer 2: The November 31 is not a real date we need to update it to November 29(Friday).

+
For example, the URI for a GET method for an endpoint named `/customers` in a particular specification might have looked like this with version 1 of the mocking service:
+
`+https://mocksvc.mulesoft.com/mocks/7d19e624-a0d6-4d77-9fb8-8c78a8aeb575/api/customers+`
+
Version 2 of the mocking service changes the URI to this and uses a different ID:
+
`+https://anypoint.mulesoft.com/mocking/api/v1/links/62a7d6af-ebc7-43fa-a815-e415e329f2ff/api/customers+`

. If any third-party tools are testing endpoints in the API specification for that project, replace the previous base URI with the new base URI in the links that those tools are using.
+
For example, suppose that this URI is being used in a tool:
+
`+https://mocksvc.mulesoft.com/mocks/7d19e624-a0d6-4d77-9fb8-8c78a8aeb575/api/customers+`
+
You would need to replace `+https://mocksvc.mulesoft.com/mocks/7d19e624-a0d6-4d77-9fb8-8c78a8aeb575+` with `+https://anypoint.mulesoft.com/mocking/api/v1/links/{new-ID}+`, where `new-ID` is the ID that version 2 of the mocking service generates.
+
You can get the new base URI from version 2 of the mocking service after step 1 above.

== New Features in Version 2 That You Can Use in API Designer

> Version 2 of the mocking service, which was released on January 10, 2019, introduced the following features that you can use in API Designer.

Reviewer 5: How about...:

After you migrate your mocking service from Version 1 to Version 2, you test OAS endpoints, add behavioral headers, enjoy improved performance, and achieve access authentication in Exchange.

I think it is important that our doc going forward follow a plan by which the text following a Level-1 heading include a statement that summarizes the Level-2 content; that a Level-2 intro text likewise summarize the Level-3 content; and so on. A user-journey roadmap, if you will, that uses signposting so that users can decide if the topic is for them or not.

=== Support for OAS 2.0

> You can now test endpoints that are in API specifications that are written in OAS 2.0.

Reviewer 5: Mocking service Version 2 enables you to test...

=== Behavioral Headers

> Add headers to your simulated calls to methods, changing the behavior of the mocking service in response to those calls The following new headers are available. See xref::apid-behavioral-headers.adoc[Add Behavioral Headers to Simulated API Calls] for full details.

Reviewer 5: Using Version 2, you can add....

> Add headers to your simulated calls to methods, changing the behavior of the mocking service in response to those calls The following new headers are available. See xref::apid-behavioral-headers.adoc[Add Behavioral Headers to Simulated API Calls] for full details.

Reviewer 5: ```suggestion
...in response to those calls. The following new headers are available:  
```

Reviewer 5: Add period, as shown above, and I suggest moving the cross-ref to follow the list rather than come between "the following..." and the actual list.

> * MS2-Delay: Indicates how many milliseconds the request should take to answer.

Reviewer 5: GLOBALLY reformat bullet list items to be one fragment or sentence, with any necessary addl' information following:

* MS2-Delay
   Indicates how many....

* MS2-Error-Level
  Indicates....

Reviewer 5: Do we mean the maximum number of milliseconds allowed for an answer, or the minimum? "How many it should take" sounds prescriptive: You should take xyz milliseconds to answer.

> * MS2-Error-Level: Indicates which range of error status codes the mocking service is restricted to use.

Reviewer 5: which range (out of an established set) or a range (any range)?

* MS2-Error-Rate: Indicates the probability of getting an error status code as a response.

> * MS2-Randomize: When this header is set to true, the mocking service auto-generates a payload response if no example is defined for the response.

Reviewer 5: set to `true`, with tagging?

Reviewer 5: auto-generates > automatically generates or autogenerates

* MS2-Status-Code: Specifies the status code or codes that the mocking service can choose from for the return message.

> * MS2-Strict-Model-Validations: When you set this header to false, you can use the mocking service even when the API specification contains errors

Reviewer 5: ditto re. "set to `<value>` " question above

Reviewer 5: End with period.

=== Performance Improvements

> The mocking service uses fewer resources, and responds to requests in less time on average.

Reviewer 5: Version 2 mocking service...resources than Version 1, and ....

Reviewer 5: Do we have to use the qualifier "on average"?

> === Authenticated Access to API Specifications That Are Shared on Exchange

Reviewer 5: Is this topic really a peer of the previous three Level-3 topics?
If so, does it perhaps contain too much detail about access authentication in this situation, maybe information that is more usefully included elsewhere (perhaps in a separate Concept)?

> Only Anypoint Platform users can simulate calls to your API specification after it is published to a private instance of Exchange, whether they are simulating calls in Exchange or using third-party tools. The mocking service now uses a base URI in Exchange that differs from the one that you can test with in API Designer.

Reviewer 5: ditto GLOBAL comment about "third-party"

Reviewer 5: Is the first sentence true for Version 1 but no longer for Version 2, or is it still true but modified by Version2?
It's unclear to me how the second sentence relates to the first; does the new base URI change the behavior indicated in the first?

> If you are in Exchange and trying to simulate a call to a method, you are prompted by Exchange to supply OAuth 2.0 credentials before sending a request.

Reviewer 5: In Version 2?

> If you are using a third-party tool to simulate a call to an Exchange asset, you must use an authorization header in each request. To get the authentication token to include in the header, you must send a `POST` request to +https://anypoint.mulesoft.com/accounts/login+ that sends the username and password for your Anypoint Platform account, as in this example.

Reviewer 5: ```suggestion
...sends the username and password for your Anypoint Platform account:
```

Reviewer 5: The colon replaces the need for ", as in this example."

```
$ curl -X POST \
  https://anypoint.mulesoft.com/accounts/login \
  -H 'Content-Type: application/json' \
  -d '{"username": "someUserName","password": "best_passw0rd"}'
```

> The mocking service responds with an authentication token, like this:

Reviewer 5: ditto: remove ", like this" and end with colon after "token".

```
{"access_token": "071e65a7-35c7-465a-8e33-ea68659099bf","token_type": "bearer","redirectUrl": "/home/"}
```

> In subsequent requests that are sent from the third-party tool, the sender must include the authentication token. For example, suppose there is the following asset in Exchange.

Reviewer 5: ```suggestion
In subsequent requests from the third-party tool, the sender must include the authentication token. For example, suppose there is the following asset in Exchange:
```

```
groupId: cb0e4f00-5026-4999-b5a5-dc2307b886c5
assetId: api-example
version: 1.0.0
```

Assume that this asset contains this API specification:

```
#%RAML 1.0
title: API Example
version: v1

/ping:
  get:
    responses:
      200:
        body:
          application/json:
            example:
              status: OK
```

A `GET` request would look like this:

```
$ curl -X GET \
  https://anypoint.mulesoft.com/mocking/api/v1/sources/exchange/assets/cb0e4f00-5026-4999-b5a5-dc2307b886c5/api-example/1.0.0/m/ping \
  -H 'MS2-Authorization: Bearer 071e65a7-35c7-465a-8e33-ea68659099bf'
```

> Assuming that the request uses the correct ID, URI, and authentication token, the mocking service would send this response:

Reviewer 5: ```suggestion
Assuming that the request uses the correct ID, URI, and authentication token, the mocking service sends this response:
```
